src/librustc_ast_passes/ast_validation.rs

```diff
@@ -883,30 +886,36 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
                 if is_auto == IsAuto::Yes {
                     // Auto traits cannot have generics, super traits nor contain items.
                     if !generics.params.is_empty() {
+                        let spans: Vec<_> = generics.params.iter().map(|i| i.ident.span).collect();
```

Comment:
I would prefer to have a single `span` for the `first.span.to(last.span)` instead of highlighting each parameter on its own.
(Also you could add a suggestion to remove all parameters in one fell swoop this way.) Since you're narrowing the span I'd also like to see a `span_label` to highlight the parameters more.

Consider adding a trait:
```rust
trait rustc_span::HasSpan {
    /// Extracts the span of the object.
    fn span(&self) -> Span;
}

impl<T: HasSpan> HasSpan for [T] {
    fn span(&self) -> Span {
        match self {
            [] => DUMMY_SP,
            [single] => single.span(),
            [first, .., last] => first.span().to(last.span()),
        }
    }
}
```

though, since we're encoding this logic ad-hoc in various places.
(Maybe we should add a `#[derive(HasSpan)]` macro for this as well, though if you fancy you can leave this to a follow up PR, in which case please make a C-cleanup issue.)
I do believe rustdoc has something like this, so we could refactor and share some code.

Reply:
I have a preference to keep the individual spans because they work better when the code is split across multiple lines. I can add a label to the last element, making the output in a single line almost indistinguishable. If the concern is around potential suggestions, we can use your strategy to create a single span, but that is a separate matter to the main output.

Reply:
Type parameters spanning across several lines are rare however, so I think using a single span optimizes for the common case (notice https://github.com/rust-lang/rust/pull/69722#discussion_r388649027 -- in the case of associated items, it would be a bad idea to use a single span because multi line is the common case.)

Adding a label to the last element doesn't feel like it makes things particularly more visible.

Reply:
I'm just failing to see how

```
error[E0568]: auto traits cannot have super traits
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:7:19
   |
LL | auto trait Magic: Copy + Clone {}
   |            -----  ^^^^^^^^^^^^
   |            |
   |            auto trait cannot have super traits
```

is better than

```
error[E0568]: auto traits cannot have super traits
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:7:19
   |
LL | auto trait Magic: Copy + Clone {}
   |            -----  ^^^^   ^^^^^
   |            |
   |            auto trait cannot have super traits
```

Reply:
The latter blames "the bound" as a collective whole, which is really the problem (`auto trait Foo: 'static` is also bad, so the diagnostic is not optimally worded here...), and facilitates highlighting the whole bound with a `.span_suggestion(span, "remove the bound", ...)`. It is the ability to have a label / suggestion that is the prize, which is diminished if you start using one label per supertrait or only highlight the last one.

Reply:
But the suggestion will already have to be a different span than the main error span, because it would otherwise have to be

```
error[E0568]: auto traits cannot have super traits
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:7:19
   |
LL | auto trait Magic: Copy + Clone {}
   |            -----^^^^^^^^^^^^^^ help: remove these super traits
   |            |
   |            auto trait cannot have super traits
```

My preferred output would be closer to

```
error[E0568]: auto traits cannot have super traits
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:7:19
   |
LL | auto trait Magic: Copy + Clone {}
   |            -----  ^^^^   ^^^^^ cannot have these super traits
   |            |
   |            auto trait cannot have super traits
help: remove the invalid super traits
   |
LL | auto trait Magic {}
   |                --

```

Reply:
I would prefer:

```rust
error[E0568]: auto traits cannot have super traits
  --> $DIR/traits-inductive-overflow-supertrait-oibit.rs:7:19
   |
LL | auto trait Magic: Copy + Clone {}
   |            -----  ^^^^^^^^^^^^ help: remove these super traits
   |            |
   |            auto trait cannot have super traits
```

This doesn't remove the `:`, but it doesn't have to, because this check is on the AST, not on the parse tree, so `auto trait Foo: {}` is perfectly legal. I think this is better than having a second `help:` below, which is more verbose.

Reply:
Also, remember that these are auto traits; they are not even stable, and are quite niche, so the amount of complexity you've expended on them now in validation does not feel proportionate.

Reply:
done